src/aurora/gff3writer.h

```diff
@@ -0,0 +1,224 @@
+ * distribution.
+ *
+ * xoreos is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 3
+ * of the License, or (at your option) any later version.
+ *
+ * xoreos is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with xoreos. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/** @file
+ *  Writer for writing version 3.x of biowares gff (general file format).
+ */
+
+#ifndef AURORA_GFF3WRITER_H
+#define AURORA_GFF3WRITER_H
+
+#include <list>
+
+#include <glm/glm.hpp>
+
+#include <boost/shared_ptr.hpp>
+
+#include "src/aurora/gff3file.h"
+#include "src/aurora/locstring.h"
+
+namespace Aurora {
+
+class GFF3WriterStruct;
+class GFF3WriterList;
+
+typedef boost::shared_ptr<GFF3WriterStruct> GFF3WriterStructPtr;
+typedef boost::shared_ptr<GFF3WriterList> GFF3WriterListPtr;
+
+class GFF3Writer {
+public:
+	//TODO: Add a constructor consuming a GFF3File object.
+	GFF3Writer(uint32 id, uint32 version = MKTAG('V', '3', '.', '2'));
+
+	/** Get the toplevel struct. */
+	GFF3WriterStructPtr getTopLevelStruct();
+
+	/** Write the gff3 to stream. */
+	void write(Common::WriteStream &stream);
+
+private:
+	/** An implementation for a field. */
+	struct Field {
+		GFF3Struct::FieldType type;
+		uint32 labelIndex;
+
+		uint32 uint32Value;
+		uint64 uint64Value;
+		int32 int32Value;
+		int64 int64Value;
+		float floatValue;
+		double doubleValue;
+		glm::vec4 vectorValue;
+		Common::UString stringValue;
+		LocString locStringValue;
+		Common::ScopedArray<byte> voidData;
+		uint32 voidSize;
+
+		Field() {
+		}
+
+		Field(const Field &field) {
+			type = field.type;
+			labelIndex = field.labelIndex;
+
+			uint32Value = field.uint32Value;
+			uint64Value = field.uint64Value;
+			int32Value = field.int32Value;
+			int64Value = field.int64Value;
+			floatValue = field.floatValue;
+			doubleValue = field.doubleValue;
+			vectorValue = field.vectorValue;
+			stringValue = field.stringValue;
+			locStringValue = field.locStringValue;
+
+			if (field.voidData.get() != 0) {
+				voidData.reset(new byte[field.voidSize]);
+				voidSize = field.voidSize;
+				memcpy(voidData.get(), field.voidData.get(), voidSize);
+			}
+		}
+
+		Field operator=(const Field &field) {
+			type = field.type;
+			labelIndex = field.labelIndex;
+
+			uint32Value = field.uint32Value;
+			uint64Value = field.uint64Value;
+			int32Value = field.int32Value;
+			int64Value = field.int64Value;
+			floatValue = field.floatValue;
+			doubleValue = field.doubleValue;
+			vectorValue = field.vectorValue;
+			stringValue = field.stringValue;
+			locStringValue = field.locStringValue;
+
+			if (field.voidData.get() != 0) {
```

Comment:
You should still do an unconditional `field.voidData.reset()` before that. Otherwise, "a = b", with a.voidData filled and b.voidData empty will not empty a.voidData.